File: course-schedule.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, vector<int>> hashmap;

        for (int i=0; i<numCourses; ++i) {
            hashmap[i] = vector<int>();
        }

        for (vector<int> pair: prerequisites) {
            hashmap[pair[0]].push_back(pair[1]);
        }

        vector<bool> visited(numCourses, false);
        for (int course=0; course < numCourses; ++course) {
            if (!finish(course, hashmap, visited)) return false;
        }
        return true;
    }

    bool finish(int course, unordered_map<int, vector<int>> hashmap, vector<bool> visited) {
        if (hashmap[course].size() == 0) return true;
        if (visited[course]) return false;

        visited[course] = true;
        for (int pre: hashmap[course]) {
            if (!finish(pre, hashmap, visited)) return false;
        }
        hashmap[course].clear();
        return true;
    }
};

File: course-schedule.py
from typing import List, Dict

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        hashmap: Dict[int, List[int]] = {i:[] for i in range(numCourses)}

        for course, pre in prerequisites:
            hashmap[course].append(pre)

        visited = [False for _ in range(numCourses)]

        def finish(course: int):
            if hashmap[course] == []:
                return True
            if visited[course]:
                return False
            
            visited[course] = True
            for pre in hashmap[course]:
                if not finish(pre):
                    return False
            hashmap[course] = []
            return True

        for course in range(numCourses):
            if not finish(course):
                return False
        return True
